### PR TITLE
Support to set result max rows for spark engine

### PR DESCRIPTION
**Pull request by turboFei**
### _Why are the changes needed?_

Support to set result max rows for spark engine.
And it will only take affect if incremental mode is not enabled.
### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by pan3793** [comment]
Can `ForcedMaxOutputRowsRule` be replaced by this one? cc @cfmcgrady @ulysses-you @wForget

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/SparkSQLOperationManager.scala [hunk 1]
```diff
@@ -64,7 +65,15 @@ class SparkSQLOperationManager private (name: String) extends OperationManager(n
             case NONE =>
               val incrementalCollect = spark.conf.getOption(OPERATION_INCREMENTAL_COLLECT.key)
                 .map(_.toBoolean).getOrElse(operationIncrementalCollectDefault)
-              new ExecuteStatement(session, statement, runAsync, queryTimeout, incrementalCollect)
+              val resultMaxRows = spark.conf.getOption(ENGINE_SPARK_MAX_ROWS.key)
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/SparkSQLOperationManager.scala:68]: move to ExecuteStatement

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 2]
```diff
@@ -626,6 +626,14 @@ object KyuubiConf {
       .intConf
       .createWithDefault(1000000)
 
+  val ENGINE_SPARK_MAX_ROWS: ConfigEntry[Int] =
+    buildConf("session.engine.spark.max.rows")
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:630]: kyuubi.operation.max.result.rows?

Comment (turboFei) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:630]: I saw that there is a parameter named `kyuubi.session.engine.flink.max.rows`.

Comment (turboFei) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:630]: How about fall back to kyuubi.operation.max.result.rows if kyuubi.session.engine.*.max.rows is not specified?

**Comment by ulysses-you** [comment]
> Can ForcedMaxOutputRowsRule be replaced by this one

It's different in physical side. `dataset.take` just like the incremental mode which run job partition by partition if previous partitions data has not satisfied the limit number. `dataset.limit.collect` will run all partitions to do local limit then do global limit in one partition. So in general `dataset.limit.collect` is more effective.

And this PR is useful. Since the limitation is controlled by client side using Incremental mode, but we can control it at engine side using `dataset.take`.

**Comment by pan3793** [comment]
> `dataset.take` just like the incremental mode which run job partition by partition if previous partitions data has not satisfied the limit number.

I would argue that `dataset.take` is same as `dataset.limit.collect`

```
def take(n: Int): Array[T] = head(n)
def head(n: Int): Array[T] = withAction("head", limit(n).queryExecution)(collectFromPlan)
def collect(): Array[T] = withAction("collect", queryExecution)(collectFromPlan)
```

**Comment by pan3793** [comment]
IIRC, the difference you described above is for increment mode.

w/ incremental mode, `CollectLimitExec#executeToIterator` will be called, which calls `getByteArrayRdd` and triggers to compute all partitions of previous stages firstly.

w/o incremental mode, `CollectLimitExec#executeCollect` will be called, which submit partitions increase exponentially utils satisfy the numbers.

**Comment by ulysses-you** [comment]
yea, if so I agree we can use this feature to replace the rule in extension.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 2 above)
Comment (turboFei) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:630]: it is to align with `kyuubi.session.engine.flink.max.rows`

**Comment by Nick-0723** [comment]
Does the user know that we truncate the data?

**Comment by turboFei** [comment]
> Does the user know that we truncate the data?

By default, it is not truncated.
It is determined by the service provider.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 2 above)
Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:630]: `kyuubi.operation.max.result.rows` makes sense to me, and we can deprecate `kyuubi.session.engine.flink.max.rows`

Comment (turboFei) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:630]: we can deprecate kyuubi.session.engine.flink.max.rows in another pr.

Comment (turboFei) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala]: resolved

File: kyuubi-server/src/test/scala/org/apache/kyuubi/operation/KyuubiOperationPerUserSuite.scala [hunk 3]
```diff
@@ -136,4 +136,21 @@ class KyuubiOperationPerUserSuite extends WithKyuubiServer with SparkQueryTests
 
     assert(r1 !== r2)
   }
+
+  test("test engine spark result max rows") {
+    withSessionConf()(Map.empty)(Map(KyuubiConf.OPERATION_RESULT_MAX_ROWS.key -> "1")) {
+      withJdbcStatement() { statement =>
```
Comment (cfmcgrady) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/operation/KyuubiOperationPerUserSuite.scala:142]: nit: `withJdbcStatement("va")`

Comment (turboFei) [reply kyuubi-server/src/test/scala/org/apache/kyuubi/operation/KyuubiOperationPerUserSuite.scala]: thanks

**Comment by cfmcgrady** [comment]
> yea, if so I agree we can use this feature to replace the rule in extension.

+1

**Review by ulysses-you — APPROVED** [review]

**Comment by ulysses-you** [comment]
thanks, merging to master

**Pull request closed without merging**